Sources/TensorFlow/Core/Runtime.swift

```diff
@@ -1281,6 +1264,32 @@ class _ThreadLocalState {
     }
 }
 
+/// Stack of devices that models nested calls to withDevice/withDefaultDevice. Devices are
+/// represented by their names in TensorFlow notation. See documentation for
+/// `withDevice(named:perform:)` to learn about device names.
+///
+/// All TensorFlow operations will be put on the topmost device on the stack. When the stack is
+/// empty or the topmost device is `nil`, that allows TensorFlow to place operations on any device
+/// that it sees fit.
+@usableFromInline
+class DeviceScopes {
+    var deviceStack: [String?] = []
+
+    var _currentDevice: String? {
+        return deviceStack.last ?? nil
+    }
+
+    @usableFromInline
+    func pushDevice(_ device: String?) {
+        deviceStack.append(device)
+    }
+
+    @usableFromInline
+    func popDevice() {
+        internalConsistencyCheck(deviceStack.popLast() != nil)
+    }
+}
+
 @usableFromInline
 @_cdecl("_swift_tfc_OpSetDeviceFromScope")
```

Comment:
I think this can be deleted now.

Reply:
This is still used in EagerExecution.swift. However, those use cases can be rewritten. Will send out a separate PR for that.

Reply:
It looks like you were only referring to `_cdecl`. Will send out a PR cleaning this up and a few other uses.

Reply:
Yeah it'd be nice to have EagerExecution.swift use C API directly.